sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java

```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction;
+
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.DONE;
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.STOP;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkState;
+
+import com.google.cloud.Timestamp;
+import java.math.BigDecimal;
+import java.util.Optional;
+import java.util.function.Supplier;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker.HasProgress;
+import org.apache.beam.sdk.transforms.splittabledofn.SplitResult;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO: Add java docs
+// TODO: Implement duration waiting for returning false on try claim
+public class PartitionRestrictionTracker
+    extends RestrictionTracker<PartitionRestriction, PartitionPosition> implements HasProgress {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PartitionRestrictionTracker.class);
+  private final PartitionRestrictionSplitter splitter;
+  private final PartitionRestrictionClaimer claimer;
+  private final PartitionRestrictionProgressChecker progressChecker;
+  protected PartitionRestriction restriction;
+  private @Nullable PartitionPosition lastClaimedPosition;
+  protected Supplier<Timestamp> timeSupplier;
+
+  public PartitionRestrictionTracker(PartitionRestriction restriction) {
+    this(
+        restriction,
+        new PartitionRestrictionSplitter(),
+        new PartitionRestrictionClaimer(),
+        new PartitionRestrictionProgressChecker());
+  }
+
+  PartitionRestrictionTracker(
+      PartitionRestriction restriction,
+      PartitionRestrictionSplitter splitter,
+      PartitionRestrictionClaimer claimer,
+      PartitionRestrictionProgressChecker progressChecker) {
+    this.splitter = splitter;
+    this.claimer = claimer;
+    this.restriction = restriction;
+    this.progressChecker = progressChecker;
+    this.timeSupplier = () -> Timestamp.now();
+  }
+
+  @VisibleForTesting
+  public void setTimeSupplier(Supplier<Timestamp> timeSupplier) {
+    this.timeSupplier = timeSupplier;
+  }
+
+  @Override
+  public @Nullable SplitResult<PartitionRestriction> trySplit(double fractionOfRemainder) {
+    final String token =
+        Optional.ofNullable(restriction.getMetadata())
+            .map(PartitionRestrictionMetadata::getPartitionToken)
+            .orElse("");
+    final SplitResult<PartitionRestriction> splitResult =
+        splitter.trySplit(fractionOfRemainder, lastClaimedPosition, restriction);
+    LOG.debug(
```

Comment:
This is already printed in the partition splitter.

Reply:
Done